File: web/src/lib/alertCopy.ts
import type { AppAlert } from '@/services/alertService'
import { categoryColor, glyphOf, goalMark } from '@/lib/categoryGlyphs'
import { budgetNote, fill, shortDate } from '@/lib/inicio'
import { shortWallet } from '@/lib/movimientos'
import type { TranslationKey } from '@/lib/i18n/translations'
import type { CategoryId, Wallet } from '@/types'
import type { MoneyTemplate } from '@/components/v2/Money'

// Web copy for each shared alert rule, per the Web v2 mockup's ALERTS
// (sentences end with a period on Web; Android's card copy doesn't). Budget
// alerts append the month's pace when it crosses the limit before month end.
// Bodies are templates with their amounts marked, so the card can blur just
// the amounts when amounts are hidden (see MoneyText).
export function alertCopy(
  alert: AppAlert,
  today: string,
  language: 'es' | 'en',
  t: (k: TranslationKey) => string,
  tCategory: (id: CategoryId) => string,
  wallets: Wallet[] = [],
): { title: string; body: MoneyTemplate[]; glyph: string[]; color: string } {
  switch (alert.kind) {
    case 'series_due':
      return {
        title: alert.overdue ? fill(t('alert.seriesOverdue.title'), alert.name, shortDate(alert.dueDate, language)) : fill(t('alert.seriesDue.title'), alert.name),
        body: [{ template: t('alert.series.body'), args: [{ amount: alert.amount }] }],
        glyph: glyphOf(alert.category),
        color: categoryColor(alert.category),
      }
    case 'loan_open':
      return {
        title: fill(t(alert.loanKind === 'lent' ? 'alert.loanLent.title' : 'alert.loanBorrowed.title'), alert.counterpartyName ?? t('loans.unknownPerson')),
        body: [{ template: t('alert.loan.body'), args: [{ amount: alert.outstanding }, shortDate(alert.dueDate, language)] }],
        glyph: glyphOf('exp.other'),
        color: categoryColor('exp.other'),
      }
    case 'budget_at_risk': {
      const note = budgetNote(alert.spent, alert.limit, alert.percentage, today)
      return {
        title: fill(t('alert.budget.title'), alert.name ?? tCategory(alert.category), alert.percentage),
        body: [
          { template: t('alert.budget.body'), args: [{ amount: alert.spent }, { amount: alert.limit }] },
          ...(note.kind === 'exceeds' ? [{ template: t('alert.budget.pace'), args: [note.days] }] : []),
        ],
        glyph: glyphOf(alert.category),
        color: categoryColor(alert.category),
      }
    }
    case 'goal_near': {
      const mark = goalMark(alert.icon)
      return {
        title: fill(t('alert.goal.title'), alert.name, alert.percentage),
        body: [{ template: t('alert.goal.body'), args: [{ amount: alert.remaining }] }],
        glyph: mark.glyph,
        color: mark.color,
      }
    }
    case 'goal_plan_due': {
      const mark = goalMark(alert.icon)
      const wallet = shortWallet(wallets.find((w) => w.id === alert.accountId)?.name ?? '')
      const when = alert.dueDate === today ? 'para hoy' : `para el ${shortDate(alert.dueDate, language)}`
      return {
        title: `Aporte programado a ${alert.name}`,
        body: [{ template: `Tienes un aporte de {0} ${alert.currency} ${when} desde {1}.`, args: [{ amount: alert.amount }, wallet] }],
        glyph: mark.glyph,
        color: mark.color,
      }
    }
    case 'goal_plan_auto': {
      const mark = goalMark(alert.icon)
      return {
        title: 'Aporte automático registrado',
        body: [{ template: `{0} ${alert.currency} a {1} desde {2} · {3}`, args: [{ amount: alert.amount }, alert.name, shortWallet(alert.walletName), shortDate(alert.date, language)] }],
        glyph: mark.glyph,
        color: mark.color,
      }
    }
    case 'tx_planned':
      return {
        title: `${alert.name || tCategory(alert.category)} se registra el ${shortDate(alert.dueDate, language)}`,
        body: [{ template: '{0} · {1} · te pediremos confirmarlo', args: [{ amount: alert.amount, currency: alert.currency }, shortWallet(alert.walletName)] }],
        glyph: glyphOf(alert.category),
        color: categoryColor(alert.category),
      }
  }
}

File: android/app/src/main/java/com/s2nova/app/ui/AlertCopy.kt
package com.s2nova.app.ui

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import com.s2nova.app.data.formatShortDate
import com.s2nova.app.ui.components.categoryColor
import com.s2nova.app.data.model.AppAlert
import com.s2nova.app.data.AppContainer
import com.s2nova.app.data.Taxonomy
import com.s2nova.app.data.fmtDate
import com.s2nova.app.data.formatMoney
import com.s2nova.app.data.model.LoanKind
import com.s2nova.app.ui.components.categoryName
import com.s2nova.app.ui.components.glyphIcon
import com.s2nova.app.ui.components.hexColor
import com.s2nova.app.ui.components.iconFor
import com.s2nova.app.ui.screens.addtransaction.shortWallet

// Where tapping an alert (Inicio card or bell row) takes the user.
enum class AlertTarget { MOVIMIENTOS, PROGRAMADOS, PLANES_BUDGETS, PLANES_GOALS, PLANES_LOANS_LENT, PLANES_LOANS_BORROWED }

data class AlertPresentation(
    val title: String,
    val body: String,
    val icon: ImageVector,
    val color: Color,
    val target: AlertTarget,
    // Set on "Aporte programado": the notification shows "Confirmar aporte"
    // / "Omitir esta vez" for this goal.
    val confirmGoalId: String? = null,
)

// Copy per S2 Nova Android v2 mockup (notification builder): the same
// wording on the Inicio card and in the bell sheet.
fun presentAlert(alert: AppAlert, t: (StringKey) -> String, format: CurrencyFormatter): AlertPresentation = when (alert) {
    is AppAlert.SeriesDue -> AlertPresentation(
        title = if (alert.overdue) {
            t(StringKey.ALERT_SERIES_OVERDUE_TITLE).format(alert.name, formatShortDate(alert.dueDate))
        } else {
            t(StringKey.ALERT_SERIES_DUE_TITLE).format(alert.name)
        },
        body = t(StringKey.ALERT_SERIES_BODY).format(format(alert.amount)),
        icon = iconFor(alert.category),
        color = categoryColor(alert.category),
        target = AlertTarget.PROGRAMADOS,
    )
    is AppAlert.LoanOpen -> {
        val person = alert.counterpartyName ?: ""
        AlertPresentation(
            title = t(if (alert.loanKind == LoanKind.LENT) StringKey.ALERT_LOAN_LENT_TITLE else StringKey.ALERT_LOAN_BORROWED_TITLE).format(person),
            body = t(StringKey.ALERT_LOAN_BODY).format(format(alert.outstanding), formatShortDate(alert.dueDate)),
            // The mockup marks loans with the "Otros" glyph in the salary green.
            icon = iconFor("exp.other"),
            color = hexColor(Taxonomy.visColor("work")),
            target = if (alert.loanKind == LoanKind.LENT) AlertTarget.PLANES_LOANS_LENT else AlertTarget.PLANES_LOANS_BORROWED,
        )
    }
    is AppAlert.BudgetAtRisk -> AlertPresentation(
        title = t(StringKey.ALERT_BUDGET_TITLE).format(alert.name ?: categoryName(alert.category), alert.percentage),
        body = t(StringKey.ALERT_BUDGET_BODY).format(format(alert.spent), format(alert.limit)),
        icon = iconFor(alert.category),
        color = categoryColor(alert.category),
        target = AlertTarget.PLANES_BUDGETS,
    )
    is AppAlert.GoalNear -> {
        val p = Taxonomy.planIcon(alert.icon)
        AlertPresentation(
            title = t(StringKey.ALERT_GOAL_TITLE).format(alert.name, alert.percentage),
            body = t(StringKey.ALERT_GOAL_BODY).format(format(alert.remaining)),
            icon = glyphIcon(p.glyph),
            color = hexColor(p.color),
            target = AlertTarget.PLANES_GOALS,
        )
    }
    // PLANS.md §3 notifications.
    is AppAlert.GoalPlanDue -> {
        val p = Taxonomy.planIcon(alert.icon)
        val wallet = AppContainer.walletRepository.wallets.value.firstOrNull { it.id == alert.walletId }?.name?.let(::shortWallet) ?: ""
        val today = java.time.LocalDate.now().toString()
        AlertPresentation(
            title = "Aporte programado a " + alert.name,
            body = "Tienes un aporte de " + formatMoney(alert.amount, alert.currency) + " " + alert.currency + " para " + (if (alert.dueDate == today) "hoy" else "el " + fmtDate(alert.dueDate)) + " desde " + wallet + ".",
            icon = glyphIcon(p.glyph),
            color = hexColor(p.color),
            target = AlertTarget.PLANES_GOALS,
            confirmGoalId = alert.goalId,
        )
    }
    is AppAlert.TxPlanned -> AlertPresentation(
        title = alert.name.ifBlank { AppContainer.categoryRepository.name(alert.category) } + " se registra el " + fmtDate(alert.dueDate),
        body = formatMoney(alert.amount, alert.currency) + " · " + shortWallet(alert.walletName) + " · te pediremos confirmarlo",
        icon = iconFor(alert.category),
        color = categoryColor(alert.category),
        target = AlertTarget.MOVIMIENTOS,
    )
    is AppAlert.GoalPlanAuto -> {
        val p = Taxonomy.planIcon(alert.icon)
        AlertPresentation(
            title = "Aporte automático registrado",
            body = formatMoney(alert.amount, alert.currency) + " " + alert.currency + " a " + alert.name + " desde " + shortWallet(alert.walletName) + " · " + fmtDate(alert.date),
            icon = glyphIcon(p.glyph),
            color = hexColor(p.color),
            target = AlertTarget.PLANES_GOALS,
        )
    }
}
